jerry-core/ecma/operations/ecma-typedarray-object.c

```diff
@@ -60,62 +51,61 @@ static ecma_number_t
 get_typedarray_element (lit_utf8_byte_t *src, /**< the location in the internal arraybuffer */
                         lit_magic_string_id_t class_id) /**< class name of the typedarray */
 {
-  ecma_number_t ret_num;
-
   switch (class_id)
   {
     case LIT_MAGIC_STRING_INT8_ARRAY_UL:
     {
-      GET_ELEMENT (int8_t, src, ret_num);
-      break;
+      return GET_ELEMENT (int8_t, src);
     }
     case LIT_MAGIC_STRING_UINT8_ARRAY_UL:
     case LIT_MAGIC_STRING_UINT8_CLAMPED_ARRAY_UL:
     {
-      GET_ELEMENT (uint8_t, src, ret_num);
-      break;
+      return GET_ELEMENT (uint8_t, src);
     }
     case LIT_MAGIC_STRING_INT16_ARRAY_UL:
     {
-      GET_ELEMENT (int16_t, src, ret_num);
-      break;
+      return GET_ELEMENT (int16_t, src);
     }
     case LIT_MAGIC_STRING_UINT16_ARRAY_UL:
     {
-      GET_ELEMENT (uint16_t, src, ret_num);
-      break;
+      return GET_ELEMENT (uint16_t, src);
     }
     case LIT_MAGIC_STRING_INT32_ARRAY_UL:
     {
-      GET_ELEMENT (int32_t, src, ret_num);
-      break;
+      return GET_ELEMENT (int32_t, src);
     }
     case LIT_MAGIC_STRING_UINT32_ARRAY_UL:
     {
-      GET_ELEMENT (uint32_t, src, ret_num);
-      break;
+      return GET_ELEMENT (uint32_t, src);
     }
     case LIT_MAGIC_STRING_FLOAT32_ARRAY_UL:
     {
-      GET_ELEMENT (float, src, ret_num);
-      break;
+      return GET_ELEMENT (float, src);
     }
 #if CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64
     case LIT_MAGIC_STRING_FLOAT64_ARRAY_UL:
     {
-      GET_ELEMENT (double, src, ret_num);
-      break;
+      return GET_ELEMENT (double, src);
     }
 #endif /* CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64 */
     default:
     {
       JERRY_UNREACHABLE ();
+      return 0;
     }
   }
-
-  return ret_num;
 } /* get_typedarray_element */
 
+#undef GET_ELEMENT
+
+/**
+ * Cast the number from ecma_numbet_t into type,
+ * and store the number in the location
+ */
+#define SET_ELEMENT(type, location, number) \
+  int32_t v = (int32_t) number; \
+  *((type *) location) = (type) v;
```

Comment:
Consider an int32 array:

https://www.ecma-international.org/ecma-262/6.0/#table-49
https://www.ecma-international.org/ecma-262/6.0/#sec-toint32

    int = floor(abs(number))
    int32bit = int mod 2^32

So if you have a double value, which is greater than 2^63, it can still have bits which affect the 32 bit part, but the conversion results with 0x80...0, so it is converted to 0.